zflow_runtime/src/go/mod.rs

```diff
@@ -0,0 +1,147 @@
+    pub ordered: bool,
+    #[serde(default)]
+    /// Whether the component should activate when it receives packets
+    pub activate_on_input: bool,
+    #[serde(default)]
+    /// Bracket forwarding rules. By default we forward
+    pub forward_brackets: HashMap<String, Vec<String>>,
+    #[serde(default = "default_base_dir")]
+    /// Base directory of lua sources
+    pub base_dir: String,
+    /// Path to lua source
+    pub source: String,
+    #[serde(default)]
+    pub package_id: String,
+    #[serde(default)]
+    pub metadata: Map<String, Value>,
+}
+
+impl GoComponent {
+    pub fn from_metadata(meta: Value) -> Option<GoComponent> {
+        GoComponent::deserialize(meta).ok()
+    }
+
+    pub fn with_metadata(&mut self, meta: Value) -> GoComponent {
+        if let Some(meta) = GoComponent::from_metadata(meta.clone()) {
+            self.inports.extend(meta.inports);
+            self.outports.extend(meta.outports);
+
+            if !meta.description.is_empty() {
+                self.description = meta.description;
+            }
+
+            if !meta.icon.is_empty() {
+                self.icon = meta.icon;
+            }
+            self.forward_brackets.extend(meta.forward_brackets);
+            if !meta.base_dir.is_empty() {
+                self.base_dir = meta.base_dir;
+            }
+        } else if let Some(meta) = meta.clone().as_object() {
+            self.metadata = meta.clone();
+        }
+        self.clone()
+    }
+}
+
+impl ModuleComponent for GoComponent {
+    fn as_component(&self) -> Result<Component, String> {
+        let node_name = self.name.clone();
+        let mut code = PathBuf::from(&self.base_dir);
+
+        let mut inports = self.inports.clone();
+        let mut outports = self.outports.clone();
+        if inports.is_empty() {
+            inports.insert("in".to_string(), PortOptions::default());
+        }
+
+        if outports.is_empty() {
+            outports.insert("out".to_string(), PortOptions::default());
+        }
+
+        return Ok(Component::new(ComponentOptions{
+            metadata: Some(self.metadata.clone()),
+            in_ports: HashMap::from_iter(
+                inports.clone()
+                    .iter()
+                    .map(|(key, options)| (
+                        key.clone(),
+                        InPort::new(options.clone())))
+                    .collect::<Vec<_>>(),
+            ),
+            out_ports: HashMap::from_iter(
+                outports.clone()
+                    .iter()
+                    .map(|(key, options)| (key.clone(), OutPort::new(options.clone())))
+                    .collect::<Vec<_>>(),
+            ),
+            description: self.description.clone(),
+            icon: self.icon.clone(),
+            ordered: self.ordered,
+            activate_on_input: self.activate_on_input,
+            forward_brackets: self.forward_brackets.clone(),
+            // question: why doesnt a component have a graph?
```

Comment:
Because if `graph` has a value pointing to a zflow subgraph, it would not need to implement the `process` function, the process function will be `None` in that case. 

Zflow's ability to trigger a subgraph is a cool feature that will allow it to run complex workflows and make all your graphs maintainable and extendable as separate modules.